modules/libcom/src/iocsh/libComRegister.c

```diff
@@ -470,6 +478,65 @@ static void installLastResortEventProviderCallFunc(const iocshArgBuf *args)
     installLastResortEventProvider();
 }
 
+/* pushd */
+static const iocshArg pushdArg0 = {"dir", iocshArgString};
+static const iocshArg* const pushdArgs[1] = {&pushdArg0};
+static const iocshFuncDef pushdFuncDef = {"pushd", 1, pushdArgs,
+                                         "Pushes the current directory to the directory stack and changes to the specified directory"};
+
+static void pushdFunc(const iocshArgBuf* args)
+{
+    char cwd[PATH_MAX+1];
+    if (!getcwd(cwd, sizeof(cwd))) {
+		perror("getcwd failed");
+		return;
+	}
+
+    if (iocshSetError(chdir(args[0].sval)) != 0) {
+        fprintf(stderr, "Invalid directory path, ignored\n");
+    }
+    else {
+        struct dirStackEntry* e = calloc(1, sizeof(struct dirStackEntry) + strnlen(cwd+1, PATH_MAX));
+        strcpy(e->path, cwd);
+        ellAdd(&dirStack, &e->node);
+    }
+}
+
+/* popd */
+static const iocshFuncDef popdFuncDef = {"popd", 0, NULL,
+                                         "Pops the top off of the directory stack and changes to that directory"};
+
+static void popdFunc(const iocshArgBuf* args)
+{
+    if (ellCount(&dirStack) <= 0) {
+        fprintf(stderr, "Directory stack is empty.\n");
+        return;
+    }
+
+    struct dirStackEntry* e = (struct dirStackEntry*)ellPop(&dirStack);
+    if (iocshSetError(chdir(e->path)) != 0) {
+        fprintf(stderr, "Invalid directory path, ignored\n");
+    }
+    free(e);
+}
+
+/* dirs */
+static const iocshFuncDef dirsFuncDef = {"dirs", 0, NULL,
+										 "Displays the contents of the directory stack, including current directory on top"};
+
+static void dirsFunc(const iocshArgBuf* args)
+{
+	struct dirStackEntry* e = NULL;
+
+	char cwd[PATH_MAX];
+	getcwd(cwd, sizeof(cwd));
```

Comment:
You need to handle the possibility of failure here.  eg. if the current directory name is very long.  Also, `PATH_MAX` may not be available on all targets.

Actually, it looks like windows supports the posix-2001 behavior for `getcwd(NULL,0)` to return a `malloc()` buffer.  So maybe something like this untested code:

```c++
std::string cwd;
#ifdef PATH_MAX
char scratch[PATH_MAX];
if(getcwd(scratch, sizeof(scratch))) {
    cwd = scratch;
} else
#endif
if(char *alloc = getcwd(NULL,0)) {
    try {
        cwd = alloc;
        free(alloc);
    } catch(std::exception&) { // only bad_alloc possible
        free(alloc);
    }
}
if(!cwd.empty())
    context->dirstack.push(cwd);
else
    iocshSetErr(1);
```

Reply:
It might be worth investigating other uses of `getcwd` as well. It seems like `PATH_MAX` is defaulted to 100 in a couple of source files, even though Windows provides `MAX_PATH` instead of `PATH_MAX`. The `cd` command uses a seemingly arbitrary buffer size of 1024 with `getcwd`.

`PATH_MAX` is defined in `limits.h` (part of the C library), and it exists in both newlib and glibc. Windows has its own `MAX_PATH` and doesn't define `PATH_MAX`. Perhaps EPICS should provide `EPICS_PATH_MAX` to consolidate these definitions and to provide a default if it's missing?

Reply:
I've added error handling to this particular call of `getcwd`, but haven't used the POSIX 2001 behavior for it. I'd like to get some feedback on the above comment before I go one way or the other.

Reply:
> It might be worth investigating other uses of `getcwd` ...

I agree, though imo. this should be a separate project.

Since these commands are a new feature, I think I am ok with not handling the long CWD case right away.